perception/sub8_perception/nodes/path_marker_finder.py

```diff
@@ -60,152 +66,283 @@ def __init__(self):
+            self.debug_pub = Image_Publisher("~debug_image")
+            self.markerPub = rospy.Publisher('~path_marker_visualization', Marker, queue_size=10)
+        self.service2D = rospy.Service('/vision/path_marker/2D', VisionRequest2D, self._vision_cb_2D)
+        if self.do_3D:
+            self.service3D = rospy.Service('/vision/path_marker/pose', VisionRequest, self._vision_cb_3D)
+        self.toggle = rospy.Service('/vision/path_marker/enable', SetBool, self._enable_cb)
+        self.image_sub = Image_Subscriber(camera, self._img_cb)
         camera_info = self.image_sub.wait_for_camera_info()
         self.cam = PinholeCameraModel()
         self.cam.fromCameraInfo(camera_info)
 
-    def sendDebugMarker(self):
+    def _send_debug_marker(self):
         '''
         Sends a rviz marker in the camera frame with the estimated pose of the path marker.
         This marker is a scaled cube with the demensions and color of the actual marker.
         Only called if debug_ros param == True
         '''
+        if self.last3d is None or not self.found:
+            return
         m = Marker()
-        m.header.frame_id = self.cam.tfFrame()
-        m.header.stamp = self.last_found_time
+        m.header.frame_id = '/map'
+        m.header.stamp = self.last_found_time_3D
         m.ns = "path_markers"
         m.id = 0
         m.type = 1
         m.action = 0
-        m.pose.position = numpy_to_point(self.last3d[0])
-        m.pose.orientation = numpy_to_quaternion(self.last3d[1])
         # Real demensions of path marker
         m.scale.x = 1.2192
         m.scale.y = 0.1524
         m.scale.z = 0.05
+        m.pose.position = numpy_to_point(self.last3d[0])
+        m.pose.orientation = numpy_to_quaternion(self.last3d[1])
         m.color.r = 0.0
         m.color.g = 0.5
         m.color.b = 0.0
         m.color.r = 1.0
         m.color.a = 1.0
-        m.lifetime = rospy.Duration(0)
+        m.lifetime = rospy.Duration(self.timeout_seconds)
         self.markerPub.publish(m)
 
-    def enable_cb(self, x):
+    def _enable_cb(self, x):
+        if x.data != self.enabled:
+            self._clear_filter(None)
         self.enabled = x.data
         return SetBoolResponse(success=True)
 
-    def sortRect(self, rect):
+    def _sort_rect(self, rect):
         '''
         Given a contour of 4 points, returns the same 4 points sorted in a known way.
         Used so that indicies of contour line up to that in model for cv2.solvePnp
-        p[0] = Top left corner of marker
-        p[1] = Top right corner of marker
-        p[2] = Bottom right corner of marker
-        p[3] = Bottom left cornet of marker
+        p[0] = Top left corner of marker         0--1
+        p[1] = Top right corner of marker        |  |
+        p[2] = Bottom right corner of marker     |  |
+        p[3] = Bottom left cornet of marker      3--2
+
+        The implementation of this is a little bit of magic, and there may be an easier way
```

Comment:
I did also did this for the tropedo board. I just took the centroid and then got 4 displacement vectors  from the centroid to the corners. Using atan2 you can get 4 angles from the vector to the positive x axis. Assuming the angles are all in [-pi, pi], it' is as easy as sorting from least to greatest.